flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java

```diff
@@ -183,6 +235,42 @@ public void notifyCheckpointComplete(long checkpointId, OperatorChain<?, ?> oper
 		env.getTaskStateManager().notifyCheckpointComplete(checkpointId);
 	}
 
+	@Override
+	public void notifyCheckpointAborted(long checkpointId, OperatorChain<?, ?> operatorChain, Supplier<Boolean> isRunning) throws Exception {
+
+		if (isRunning.get()) {
+			LOG.debug("Notification of aborted checkpoint for task {}", taskName);
+			// only happens when the task always received checkpoints to abort but never trigger or executing.
+			if (abortedCheckpointIds.size() >= maxRecordAbortedCheckpoints) {
+				abortedCheckpointIds.pollFirst();
+			}
+
+			channelStateWriter.abort(checkpointId, new NotifiedCheckpointAbortedException(checkpointId));
+			boolean canceled = asyncCheckpointRunnableRegistry.cancelAsyncCheckpointRunnable(checkpointId);
+
+			if (!canceled) {
+				if (checkpointId > lastCheckpointId) {
+					// only record checkpoints that have not triggered on task side.
+					abortedCheckpointIds.add(checkpointId);
+				}
+			}
+
+			for (StreamOperatorWrapper<?, ?> operatorWrapper : operatorChain.getAllOperators(true)) {
+				operatorWrapper.getStreamOperator().notifyCheckpointAborted(checkpointId);
```

Comment:
What if `notifyCheckpointAborted` fails for one operator, should we continue to the next?

Reply:
I think this is the same for `notifyCheckpointComplete`, catch the exception and continue to the next operator should be a solution.

Reply:
For `notifyCheckpointComplete` it's less important because operators are expected to work correctly even without the notification (and `channelStateWriter` has only in-memory resources).

Another question: this is a new method and we call it only on RPC notification, right? Should we also call it in `abortCheckpointOnBarrier`?